Implemented Models for Transactions, Users, Banks, and Bank Accounts

Hamza-Chaudhary65 opened the pull request:
This PR includes the initial implementation of the models for Fintrack .

Apps:  banks, transactions, users, bank_accounts

1. `User`: Extends the built-in Django Abstract User model to include additional profile information.
2. `User Manager`
3. `Bank`: Represents a bank with a unique name.
4. `BankBranch`: Represents a branch of a bank, including the branch name and branch code.
5. `BankAccount`: Represents a bank account linked to a user and a branch, including account number, holder name, account type, and current balance.
6. `Category` and `SubCategory`: Represent the categories  for transactions and subcategories depends on categories.
7. `Transaction`: Represents a financial transaction, linked to a bank account and category.
8. `Transaction Helper`:  contains logic for managing accounts after any successful transaction
9.  `populate_data.py`:  just for testing database models and their relations.

Hamza-Chaudhary65 committed:
created models for banks app

Hamza-Chaudhary65 committed:
created models for apps : banks, transactions, bank_accounts, users.
created populate_data for testing models and relations.
created handler in transactions to manage logics related to transactions

Hamza-Chaudhary65 committed:
eof

sheralim012 commented on fintrack/bank_accounts/models.py:
```diff
@@ -0,0 +1,22 @@
+
+
+class BankAccount(models.Model):
+    class AccountTypeChoices(models.TextChoices):
```
create a choices.py file and move it there
it will allow you to easily access choices in other files

sheralim012 commented on fintrack/bank_accounts/__init__.py:
The project is related to banks and accounts will be of the banks so no need to add prefix of bank
rename the app to accounts

sheralim012 commented on fintrack/bank_accounts/models.py:
```diff
@@ -0,0 +1,22 @@
+        Saving = "Saving", _("Saving")
+
+
+    acc_holder_name = models.CharField(max_length=100, blank=False)
+    acc_number = models.CharField(max_length=20, primary_key=True)
+    acc_type = models.CharField(choices=AccountTypeChoices.choices, max_length=100, blank=False)
```
only use prefix when context isn't available

sheralim012 commented on fintrack/banks/models.py:
```diff
@@ -0,0 +1,22 @@
+
+class BankBranch(models.Model):
+    branch_name = models.CharField(max_length=100)
+    branch_code = models.CharField(max_length=100, unique=True)
```
add address and phone

sheralim012 commented on fintrack/transactions/models.py:
```diff
@@ -0,0 +1,30 @@
+from users.models import User
+
+
+class Category(models.Model):
+    name = models.CharField(max_length=50, blank=False, unique=True)
+
+    def __str__(self):
+        return self.name
+
+
+class SubCategory(models.Model):
+    sub_category = models.CharField(max_length=50, blank=False, unique=True)
+
+    category = models.ForeignKey(Category, on_delete=models.CASCADE)
+
+    def __str__(self):
+        return self.sub_category
```
we can avoid using two models?

sheralim012 commented on fintrack/transactions/models.py:
```diff
@@ -0,0 +1,30 @@
+
+    amount = models.PositiveIntegerField(blank=False)
+
+    transaction_from_account = models.ForeignKey("bank_accounts.BankAccount", on_delete=models.CASCADE)
```
should we add destination account as well?

Hamza-Chaudhary65 replied:
Sir I considered it, but transaction subcategories are like food, transport, groceries so I think it doesn't needs destination account

sheralim012 commented on fintrack/users/models.py:
```diff
@@ -0,0 +1,10 @@
+
+
+class User(AbstractUser):
+    address = models.CharField(max_length=250)
```
do we need email and username separate or one should be enough?

Hamza-Chaudhary65 replied:
Username is for login purposes as it is short and handy. where as we are taking email for marketing or reset purposes. if user forgot his username or password, an email could be sent to his registered email to reset it.

sheralim012 commented on fintrack/users/user_manager.py:
```diff
@@ -0,0 +1,18 @@
+from django.contrib.auth.base_user import BaseUserManager
+
+
+class UserManager(BaseUserManager):
```
isn't it the default user manager?

sheralim012 commented on fintrack/transactions/transaction_helper.py:
```diff
@@ -0,0 +1,32 @@
+from bank_accounts.models import BankAccount
+
+
+def create_transaction(description, date, amount, account, category, sub_cat):
```
I would suggest to create a utility/helper class instead

sheralim012 commented on fintrack/transactions/transaction_helper.py:
```diff
@@ -0,0 +1,32 @@
+
+def delete_transaction(transaction):
+    bank_acc = transaction.transaction_from_account
+    if transaction.sub_category.category == "Expense":
```
instead of using string literals always use constants

Hamza-Chaudhary65 committed:
Update models.
seprate classes for handlers
used constants instead of string literals
moved choices to separate file

Hamza-Chaudhary65 committed:
eof

sheralim012 commented on fintrack/banks/models.py:
```diff
@@ -0,0 +1,24 @@
+        return self.name
+    
+
+class BankBranch(models.Model):
```
```suggestion
class Branch(models.Model):
```

don't repeat mistakes, you can make new mistakes but no room for repeating mistakes

sheralim012 commented on fintrack/banks/models.py:
```diff
@@ -0,0 +1,24 @@
+        return f"{self.name} ({self.code})"
+    
+    @property
+    def bank_name(self):
```
is this necessary?

sheralim012 commented on fintrack/transactions/models.py:
```diff
@@ -0,0 +1,17 @@
+class Transaction(models.Model):
+
+    description = models.CharField(max_length=300, blank=True)
+    category = models.CharField(max_length=20, choices=TransactionTypeChoices.choices, blank=False)
```
either use category or type not both, while writing code make sure to be consistent

sheralim012 commented on fintrack/transactions/models.py:
```diff
@@ -0,0 +1,17 @@
+    description = models.CharField(max_length=300, blank=True)
+    category = models.CharField(max_length=20, choices=TransactionTypeChoices.choices, blank=False)
+
+    date = models.DateField(blank=False)
```
isn't default value for blank False?

Hamza-Chaudhary65 committed:
Model names changed
removed unncessary functions
